pkg/bucketrequest/bucketrequest.go

```diff
@@ -0,0 +1,174 @@
+package bucketrequest
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/kubernetes-sigs/container-object-storage-interface-api/apis/objectstorage.k8s.io/v1alpha1"
+	bucketclientset "github.com/kubernetes-sigs/container-object-storage-interface-api/clientset"
+	bucketcontroller "github.com/kubernetes-sigs/container-object-storage-interface-api/controller"
+	"github.com/kubernetes-sigs/container-object-storage-interface-controller/pkg/util"
+	kubeclientset "k8s.io/client-go/kubernetes"
+
+	"github.com/golang/glog"
+)
+
+type bucketRequestListener struct {
+	kubeClient   kubeclientset.Interface
+	bucketClient bucketclientset.Interface
+}
+
+func NewListener() bucketcontroller.BucketRequestListener {
+	return &bucketRequestListener{}
+}
+
+func (b *bucketRequestListener) InitializeKubeClient(k kubeclientset.Interface) {
+	b.kubeClient = k
+}
+
+func (b *bucketRequestListener) InitializeBucketClient(bc bucketclientset.Interface) {
+	b.bucketClient = bc
+}
+
+// Add creates a bucket in response to a bucketrequest
+func (b *bucketRequestListener) Add(ctx context.Context, obj *v1alpha1.BucketRequest) error {
+	glog.V(1).Infof("add called for bucket %s", obj.Name)
+	bucketRequest := obj
+	status, err := b.provisionBucketRequestOperation(ctx, bucketRequest)
+	if err == nil || status == "Finished" {
+		// Provisioning is 100% finished / not in progress.
+		switch err {
+		case nil:
+			glog.V(5).Infof("BucketRequest processing succeeded, removing bucketRequest %s from bucketRequests in progress", bucketRequest.UID)
+		case util.ErrStopProvision:
+			glog.V(5).Infof("Stop provisioning, removing bucketRequest %s from bucketRequests in progress", bucketRequest.UID)
+			// Our caller would requeue if we pass on this special error; return nil instead.
+			err = nil
+		default:
+			glog.V(2).Infof("Final error received, removing buckerRequest %s from bucketRequests in progress", bucketRequest.UID)
+		}
+		return err
+	}
+	if status == "InBackground" {
+		//nothing for now
+	} else {
+		// status == ProvisioningNoChange.
+		// Don't change bucketRequestsInProgress:
+		// - the bucketRequest is already there if previous status was ProvisioningInBackground.
+		// - the bucketRequest is not there if if previous status was "Finished".
+	}
+	return nil
```

Comment:
What if `err != nil` and `status != "Finished"`?

Reply:
Reworked this code to return and process just error.